File: nodejs/test/response-contract.test.ts
/**
 * `src/response-contract.json` restates, in the token vocabulary the live smoke
 * test speaks, what `src/types.ts` declares. Two files saying the same thing can
 * disagree, and a contract that has drifted from the types it claims to mirror
 * would check a live deployment against a promise the package no longer makes.
 *
 * These tests hold them together: same fields, and every token justified by the
 * declared type. Adding a field to `Nonprofit` fails here until the contract
 * learns about it, which is the point — a new field is a new prediction.
 */
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import ts from 'typescript';
import { describe, expect, it } from 'vitest';

interface ResponseContract {
  note: string;
  envelope: Record<string, string>;
  errorDetail: Record<string, string>;
  nonprofit: Record<string, string>;
  organizationType: Record<string, string>;
}

const contract = JSON.parse(
  readFileSync(fileURLToPath(new URL('../src/response-contract.json', import.meta.url)), 'utf8'),
) as ResponseContract;

const typesPath = fileURLToPath(new URL('../src/types.ts', import.meta.url));

const source = ts.createSourceFile(
  typesPath,
  readFileSync(typesPath, 'utf8'),
  ts.ScriptTarget.ES2022,
  true,
);

/**
 * Declared members of an interface, as field name to the source text of its
 * type. The index signature is skipped: `[key: string]: unknown` is what makes
 * an unpredicted field readable, not a field the package predicts.
 */
function declaredMembers(name: string): Map<string, string> {
  const members = new Map<string, string>();

  for (const statement of source.statements) {
    if (!ts.isInterfaceDeclaration(statement) || statement.name.text !== name) {
      continue;
    }

    for (const member of statement.members) {
      if (ts.isPropertySignature(member) && member.type && ts.isIdentifier(member.name)) {
        members.set(member.name.text, member.type.getText(source));
      }
    }
  }

  expect(members.size, `interface ${name} was not found in types.ts`).toBeGreaterThan(0);

  return members;
}

/** The token kinds a declared TypeScript type permits. */
function permittedBy(declared: string): Set<string> {
  const permitted = new Set<string>();

  for (const part of declared.split('|').map(one => one.trim())) {
    if (part === 'null') {
      permitted.add('null');
    } else if (part === 'boolean' || part === 'number') {
      permitted.add(part);
    } else if (part === 'string') {
      permitted.add('string');
    } else if (part.endsWith('[]')) {
      permitted.add('array');
    } else {
      permitted.add('object');
    }
  }

  return permitted;
}

const STRING_TOKENS = new Set(['string', 'text', 'date', 'date:iso', 'url', 'ofac-sentence', 'empty']);

/** Whether one contract token is justified by the declared type. */
function justified(token: string, permitted: Set<string>): boolean {
  if (STRING_TOKENS.has(token) || token.startsWith('digits:')) {
    return permitted.has('string');
  }

  return permitted.has(token);
}

const SHAPES: ReadonlyArray<[keyof ResponseContract & string, string]> = [
  ['envelope', 'ApiEnvelope'],
  ['errorDetail', 'ApiErrorDetail'],
  ['nonprofit', 'Nonprofit'],
  ['organizationType', 'OrganizationType'],
];

describe('response contract', () => {
  it.each(SHAPES)('%s predicts exactly the fields %s declares', (shape, interfaceName) => {
    const declared = [...declaredMembers(interfaceName).keys()]
      // `data` is the envelope's generic slot; each endpoint fills it, so the
      // contract supplies it at composition rather than declaring it once here.
      .filter(field => !(interfaceName === 'ApiEnvelope' && field === 'data'))
      .sort();

    expect(Object.keys(contract[shape] as Record<string, string>).sort()).toEqual(declared);
  });

  it.each(SHAPES)('%s tokens are all justified by %s', (shape, interfaceName) => {
    const declared = declaredMembers(interfaceName);

    for (const [field, allowed] of Object.entries(contract[shape] as Record<string, string>)) {
      const type = declared.get(field);

      expect(type, `${field} is in the contract but not in ${interfaceName}`).toBeDefined();

      const permitted = permittedBy(type as string);

      for (const token of allowed.split('|')) {
        expect(
          justified(token, permitted),
          `${interfaceName}.${field} is declared \`${type}\`, which permits no ${token} value`,
        ).toBe(true);
      }
    }
  });

  it('marks a field nullable exactly when the declared type does', () => {
    for (const [shape, interfaceName] of SHAPES) {
      const declared = declaredMembers(interfaceName);

      for (const [field, allowed] of Object.entries(contract[shape] as Record<string, string>)) {
        const type = declared.get(field) as string;

        expect(
          allowed.split('|').includes('null'),
          `${interfaceName}.${field} is declared \`${type}\` but the contract says \`${allowed}\``,
        ).toBe(permittedBy(type).has('null'));
      }
    }
  });
});

File: ruby/test/response_contract_test.rb
# frozen_string_literal: true

require_relative "test_helper"

# `response_contract.json` restates, in the token vocabulary the live smoke test
# speaks, what the field declarations in `models.rb` say. Two files saying the
# same thing can disagree, and a contract that has drifted from the models it
# claims to mirror would check a live deployment against a promise the gem no
# longer makes.
#
# These tests hold them together: same fields, and every token justified by the
# declared type. Declaring a field on `Nonprofit` fails here until the contract
# learns about it, which is the point — a new field is a new prediction.
class ResponseContractTest < Minitest::Test
  include TestSupport

  CONTRACT = JSON.parse(File.read(File.expand_path("../lib/pactman/nonprofit_check_plus/response_contract.json",
                                                   __dir__)))

  # Each contract section and the model it restates.
  SHAPES = {
    "envelope" => NCP::ApiEnvelope,
    "errorDetail" => NCP::ApiErrorDetail,
    "nonprofit" => NCP::Nonprofit,
    "organizationType" => NCP::OrganizationType
  }.freeze

  STRING_TOKENS = %w[string text date date:iso url ofac-sentence empty].freeze

  # The token kinds a field declaration permits.
  def permitted_by(field)
    kinds = field.types.map do |type|
      if type == String then "string"
      elsif type == :boolean then "boolean"
      elsif type.is_a?(Array) || type == Array then "array"
      elsif type.is_a?(Class) && type <= Numeric then "number"
      elsif type == Hash || (type.is_a?(Class) && type < NCP::Model) then "object"
      else flunk "#{field.name} declares a type the contract has no token for: #{type.inspect}"
      end
    end

    kinds << "null" if field.nullable
    kinds
  end

  def justified?(token, permitted)
    return permitted.include?("string") if STRING_TOKENS.include?(token) || token.start_with?("digits:")

    permitted.include?(token)
  end

  def test_each_section_predicts_exactly_the_fields_its_model_declares
    SHAPES.each do |section, model|
      refute_empty model.fields, "#{model} declares no fields"
      assert_equal model.fields.keys.sort, CONTRACT.fetch(section).keys.sort, "#{section} vs #{model}"
    end
  end

  def test_every_token_is_justified_by_the_declared_type
    SHAPES.each do |section, model|
      CONTRACT.fetch(section).each do |name, allowed|
        field = model.fields.fetch(name)
        permitted = permitted_by(field)

        allowed.split("|").each do |token|
          assert justified?(token, permitted),
                 "#{model}.#{name} is declared #{field.types.inspect} (null: #{field.nullable}), " \
                 "which permits no #{token} value"
        end
      end
    end
  end

  def test_marks_a_field_nullable_exactly_when_the_model_does
    SHAPES.each do |section, model|
      CONTRACT.fetch(section).each do |name, allowed|
        assert_equal model.fields.fetch(name).nullable, allowed.split("|").include?("null"),
                     "#{model}.#{name} is declared null: #{model.fields.fetch(name).nullable} " \
                     "but the contract says #{allowed}"
      end
    end
  end

  def test_requires_exactly_the_fields_the_model_declares_required
    SHAPES.each do |section, model|
      declared = model.fields.values.select(&:required).map(&:name).sort

      assert_equal declared, CONTRACT.fetch("required").fetch(section, []).sort, section
    end
  end

  def test_every_declared_field_has_a_reader_that_returns_the_wire_value
    SHAPES.each_value do |model|
      model.fields.each_value do |field|
        instance = model.new({ field.name => "wire value" })

        assert_equal "wire value", instance.public_send(field.reader), "#{model}##{field.reader}"
      end
    end
  end
end
